### DISCUSSION TITLE
Determining Datafetcher by inline fragement in the query at runtime.

### DISCUSSION THREAD

**Discussion by PoemDaga** [graphql-java/graphql-java#2246] [Q&A]
**Is there a way to determine which datafetcher will be used based on Inline Fragments in the query?**

In our query we have three main types: Event, Task, Action that has three data fetchers respectively -> EventDF, TaskDF and Action DF.

Example query:
- ETA: `query lookupQuark(){ lookupQuark(){ ... on Event{ id type child{... on Task{name type child{... on Action{name type }}}}}}}`
- ET: `query lookupQuark(){ lookupQuark(){ ... on Event {id type child{... on Task{name type }}}}}`

**We want to support different _quark_ patterns like _ETA(Event-> events children as Task ->then Action), EA, ET, TT_  which will be determined at runtime**

Question: How to build the Runtime Wiring Class?

PFA for schema
[graphql-schema.txt](https://github.com/graphql-java/graphql-java/files/6161337/graphql-schema.txt)

**Comment by bbakerman** [comment]
Fragments are not a factor that lead to the selection of DataFetchers - fields are - there is one data fetcher for every "object type + field name" combination.

Graphql goes through a complicated process called "field merging" to come up with the fields to be executed

See https://www.graphql-java.com/blog/deep-dive-merged-fields/ for more details

I suspect what you are asking is - how do I know in `lookupQuark` what POSSIBLE fields and object types might be asked for?

That is you are looking at the field sub selection.

See https://www.graphql-java.com/blog/deep-dive-data-fetcher-results/ for more details

In graphql-java 16 we have introduced a even more powerful sub selection system that allows you to KNOW possible fields present.

See `graphql.schema.DataFetchingFieldSelectionSet` and `graphql.schema.SelectedField` for more details on getting a look ahead on what possible types could be invoked

**Reply by PoemDaga** [reply]
@bbakerman That is right, 
Ask is: How do I know in lookupQuark what POSSIBLE fields and OBJECT TYPES might be asked for?

Now, thinking of a design where EACH TYPE has a DATAFETCHER (as data is specific to type). But this is not possible as per: https://github.com/graphql-java/graphql-java/issues/11

Looking ahead via DataFetchingFieldSelectionSet https://www.graphql-java.com/blog/deep-dive-data-fetcher-results/ can help in usecases where there is 1-2 level of nesting. _In our usecase, Pattern can be ETA, ETATE, EA, EE infact._

Using this feature in our usecase will add more complexity
Ex:
In Quark DF check if Parent type is
- Event
    - if inline fragment is Task 
          - if inline fragement is Action...  ->then lookupEventWithTaskAndAction()
          - else if Event
          - ......
    - else if inline fragement is Action -> then then lookupEventWithAction()
    - else (Event) ->
- Task
    - ....
- Action
    - ....

 (_It is more of a recursive problem_)

@bbakerman Correct me if I am wrong here:  "Inline Fragements" or "object type" cannot be tied to fetchers, therefore we cannot completely rely on graphql runtime wiring to determine executing flow for a generic query like `lookupQuark`

**Comment by bbakerman** [comment]
> Correct me if I am wrong here: "Inline Fragements" or "object type"
cannot be tied to fetchers

you are right - you cant work at the AST level (the parsed query) to work
out possible selection

However the latest version of graphql-java has enhanced
DataFetchingFieldSelectionSet so that it builds what we called a
"normalised tree" of fields INCLUDING possible fields (for interfaces and
union sub fields)

We havent updated the doco on it but you can do things like

```
List<SelectedField> fields =
graphql.schema.DataFetchingFieldSelectionSet#getFields()
List<SelectedField> immediateFields =
graphql.schema.DataFetchingFieldSelectionSet#getImmediateFields

// we also have glob pattern access
List<SelectedField> fields =
graphql.schema.DataFetchingFieldSelectionSet().getFields(globPattern)


SelectedField sf = ...

sf.isConditional()   /// interface and unions might NOT result in this
field;
sf.getParentField() /// its a tree
sf.getSelectionSet()   // child sub selections

```

In short this gives you the raw tool set to find ALL possible sub
selections from the point of view of the current fettching field.

You have to know your field sub structure of course to know about peeking
ahead

**Reply by bbakerman** [reply]
```
query lookupQuark(){ lookupQuark(){ ... on Event{ id type child{... on Task{name type child{... on Action{name type }}}}}}}
```

With the new `DataFetchingFieldSelectionSet` enhancements you will have a 3 fields called `type` and they will have `conditional=true` (eg via the fragments it MAY be present) and its object types of each will be `Event`, `Task`, `Action`
